src/org/jitsi/impl/neomedia/jmfext/media/protocol/audiosilence/DataSource.java

```diff
@@ -386,11 +386,27 @@ public synchronized void stop()
 
             boolean interrupted = false;
 
+            // a workaround for an issue we see where we cannot stop this stream
+            // as the thread waiting to transfer data is waiting for data that
+            // never comes, so we timeout after short period and we interrupt
+            // the thread to clean it
+            long WAIT_TIMEOUT = 100; // ms.
+            boolean waited = false;
             while (thread != null)
             {
+                if (waited)
+                {
+                    // our stop had timed out, so let's interrupt the thread
+                    this.thread.interrupt();
+                    break;
+                }
+
                 try
                 {
-                    wait();
+                    long started = System.currentTimeMillis();
+                    wait(WAIT_TIMEOUT);
+                    waited
+                        = System.currentTimeMillis() - started >= WAIT_TIMEOUT;
```

Comment:
This requires a single `wait` to last at least 100ms. I think you can just move the `started` initialization above the loop.

Reply:
@damencho 
`currentTimeMillis` is not guaranteed to be monotonic. If you run this and the time is adjusted by the OS/VM host, you might get a negative delta, possibly in the range of hours. Better use [System.nanoTime()](https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#nanoTime--) instead, and be aware of the possible numeric overflow mentioned in the Javadoc.

Reply:
@ibauersachs Something like that: https://github.com/jitsi/libjitsi/pull/530?